man/metaserver.1
.TH METASERVER SECTION "July 23, 2002"
.SH NAME
metaserver \- A simple napster metaserver. Redirects clients to a specific set of servers.
.SH SYNOPSIS
.B metaserver
.RI [ -v ] [ -l IP ] [ -p <port> ] [ host:port ... ]
.br
.SH DESCRIPTION
This manual page documents briefly the
.B metaserver
command.
This manual page was written for the Debian GNU/Linux distribution
because the original program does not have a manual page (yet).
.PP
.SH OPTIONS
A summary of options is included below.
.TP
.B \-h
Show summary of options.
.TP
.B \-v
Show version of program.
.TP
.B \-l
IP listen only on interface for IP.
.TP
.B \-p <port>
Listen for connection on <port> (default is 8875).
.SH SEE ALSO
.BR opennap (1).

.SH AUTHOR
This manual page was written by Johan Schurer <leodav@users.sf.net>,
for the Debian GNU/Linux system (but may be used by others).


NEWS
$Id: NEWS 297 2004-08-29 21:14:20Z khaytsus $

Open Souce Napster Server
-------------------------
[opennap-ng 0.48]

Config dirs have changed, you WILL POSSIBLY NEED TO MOVE FILES!  The new
structure is as such:

$prefix/etc/opennap-ng is the new Config Directory, and the following
files are within in:
class, config, filter, block, motd, servers

$prefix/var/opennap-ng is the new State Directory, and the following
files are within it:
bans, channels, opennap-ng.pid, log, level.log, users

docs/scripts/fixfiles is a script to fix this situation for you, or
you may do it yourself.

[opennap-ng 0.47]

Lots of bug fixes, some day I'll get a decent list of bug fixes and new
features.

Added some patches by Hoshi that should improve the servers latency,
speed up searches and logoff hash removals.

[opennap-ng 0.46]

This file highlights the major user visible changes between versions.  For a
detailed description of all changes, please see the ChangeLog file.

[opennap 0.44]

Fixed to speed up the process of removing files from the search index so
that there is no longer a noticeable pause when a user with a large
collection disconnects from the server.

[opennap 0.43]

Added `eject_after' variable to specify how long a client is exempt from
getting killed to make room for other clients after initial login for
not sharing enough files (default: 120 seconds).

Added `index_path_depth' variable to control how many levels of directory
are indexed for searching.  Often the leading parts of the path aren't
useful for searches, so indexing just the last 2 levels provides enough
info without wasting a lot of memory.  (Default: 2)

Added `index_ignore_suffix' variable to control whether filename extensions
are indexed.  (Default: yes)

[opennap 0.42]

stat_server_pass and stat_server_user are now hidden variables, meaning that
they can't be queried via the server.

added -r command line option, and remote_config boolean option to disable
remote configuration of the server.  only the rehash function will cause the
existing configuration to change.  none of the server variables can be
queried or changed when in this mode.

you should no longer need to double quote values in the config file if they
contain spaces (this was a bug in 0.41)

SIGHUP will now cause opennap to re-read its config file

Included registry entry for Windows 98SE to increase the max number of
allowed connectsion up to 1024.  A BIG THANKS to Damian Hodgkiss
<mian@thirty4.com> of the Napigator project for this gem of information.
See the FAQ for more details.

The Win32 port is once again compiled natively using MS Visual C++ instead of
the CYGWIN toolkit.

Fixed crash in search_result_end which could be triggered by a user.
Missing a call to check for a server issuing the command.

[opennap 0.41]

Improved Napigator support.  Added new variables report_name, report_ip, and
report_port which set the information reported to Napigator.  These default
from the other configuration variables and only need to be set if OpenNap
doesn't guess the correct values.

[opennap 0.40]

Added a `block' configuration file which contains a list of regular
expressions.  Any file a client attempts to share that matches one of the
regexps will be discarded and unavailable for download via the server.  See
sample.block for a useful list to begin with.

Added `allow_share' boolean configuration options which controls whether or
not clients are allowed to share files.  Requests to share files are
silently ignored.

Added `min_read' configuration variable to allow a low-watermark log
message when less than this many bytes are read from a server link.  Useful
for detecting slow servers.

Added support for directly reporting server stats to a Napigator-style
server.

[opennap 0.39]

The metaserver can now read the servers list from a file as well as from the
command line.

Added `server_chunk' config variable to control the minimum packet size to
be sent between servers (Default: 0 = no minimum).

Added `ghost_kill' boolean to allow disabling the ghost kill feature.

Added `auto_link' boolean to make the server automatically attempt to link
to all servers upon startup.

Added support for ircd-style i:lines (allowed ips), e:lines (exempt ips),
d:lines (drop without any notification)

Added `login_interval' and `register_interval' variables to control how fast
a client from the same ip may log in or register a nickname.

Support for Napster BETA 8 and BETA 9.

Added `remote_browse' boolean variable to disable remote browsing.

Added support for histogram of number and bytecount for each message type
received.

Disabled the 10301 (`new' browse method) [Note: a few clients depended on
this, but should have been updated to use BETA8+ method of direct client to
client browsing.]

Added `max_searches' variable to limit the number of pending searches a
client can have.

Added usermode -msg to disable private messages.

Added usermode -whois to disable whois notifications for mod+ users.

[opennap 0.38]

***NOTE*** Because of a change in the server quit messages, if 0.38 is used
as the hub it will likely crash leaf servers which are not also 0.38.  It
should be fine to use 0.38 as a leaf server with 0.37 as the hub server
until all other servers are upgraded.

Added new command `who was' (10121) to allow a mod+ to see whois-like info
about a user that has recently logged off.  Information about clients is
cached for `who_was_time' seconds before being purged.  By default, this is 5
minutes.

`max_clones' now works globally instead of per-server.

added `max_searches' config variable which controls how many pending
searches a client may queue up (default: 3)

added --disable-kqueue to turn off kqueue() support under FreeBSD

added new mass kill by ip command (10122).

channel prefixed with & are now local to a single server, even when linked.

added --enable-log-channel which creates a channel &LOG in which the console
output is displayed.


src/metaserver.c
/* Copyright (C) 2000-1 drscholl@users.sourceforge.net
This is free software distributed under the terms of the
GNU Public License.  See the file COPYING for details.

$Id: metaserver.c 430 2006-07-29 20:22:19Z reech $ */

/* a simple napster metaserver.  redirects clients to a specific set of
servers */

#if HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <limits.h>

#ifdef WIN32
# include "win32-support.h"
# define strdup _strdup
#else
# include <signal.h>
# include <unistd.h>
# include <sys/socket.h>
# include <netinet/in.h>
# include <arpa/inet.h>
# define WRITE write
# define CLOSE close
#endif /* WIN32 */

#if HAVE_POLL
#include <sys/poll.h>
#endif /* HAVE_POLL */

#if HAVE_SYSLOG
#include <syslog.h>
#endif

char    metafile[_POSIX_PATH_MAX];
char   *hosts[64];
int     numhosts = 0;
volatile int sig_meta = 0;

#ifndef HAVE_SOCKLEN_T
typedef unsigned int socklen_t;
#endif


#ifndef WIN32
static void
handler (int sig)
{
    if (sig == SIGHUP)
        sig_meta++;
}
#endif

static void
usage (void)
{
    puts ("usage: metaserver [ -fsv ] [ -c CONFIG ] [ -l IP ] [ -p <port> ] [ host:port ... ]");
    puts ("  -c CONFIG  use list of servers in file CONFIG");
    puts ("  -f         run the metaserver in the background");
    puts ("  -v         display version number and exit");
    puts ("  -l IP      listen only on interface for IP");
    puts ("  -p <port>  listen for connection on <port> (default is 8875)");
    puts ("  -s         disable syslog messages");
    puts ("\n  if no arguments are given, defaults to 127.0.0.1:8888");
    exit (1);
}

static void
read_metafile (char *filename)
{
    char    buffer[90];

    FILE   *fp;

    fp = fopen (filename, "r");
    if (!fp)
        return;
    while (fgets (buffer, sizeof (buffer) - 1, fp))
    {
        buffer[strlen (buffer) - 1] = 0;
        if (strlen (buffer) > 0 && buffer[0] != '#')
            hosts[numhosts++] = strdup (buffer);
    }
    fclose (fp);
}

int
main (int argc, char **argv)
{
    struct sockaddr_in sin;
    int     i, s, f, port = 8875;
    int     location = 0;
    int     sys_log = 1;
    int     background = 0;

#if HAVE_POLL
    struct pollfd ufd;
#else
    fd_set  set;
#endif /* HAVE_POLL */
    socklen_t sinsize;

#ifndef WIN32
    struct sigaction sa;
#endif
    unsigned int iface = INADDR_ANY;

#ifdef WIN32
    WSADATA wsa;

    WSAStartup (MAKEWORD (1, 1), &wsa);
#endif /* !WIN32 */

    metafile[0] = 0;

    while ((i = getopt (argc, argv, "hfl:vp:c:s")) != EOF)
    {
        switch (i)
        {
        case 'l':
            iface = inet_addr (optarg);
            break;
        case 'p':
            port = atoi (optarg);
            break;
        case 'c':
            strcpy (metafile, optarg);
            break;
        case 'f':
            background ^= 1;
            break;
        case 's':
            sys_log ^= 1;
            break;
        case 'v':
            printf ("%s metaserver version %s\n", PACKAGE, VERSION);
            exit (1);
        default:
            usage ();
        }
    }

    /* read in the host list */
    if (!metafile[0])
    {
        if (!argv[optind])
            hosts[numhosts++] = strdup ("127.0.0.1:8888");  /* use default host */
        else
        {
            while (argv[optind])
            {
                hosts[numhosts++] = strdup (argv[optind]);
                optind++;
            }
        }
    }
    else
        read_metafile (metafile);

#ifndef WIN32
    /* set some signal handlers so we can shut down gracefully */
    memset (&sa, 0, sizeof (sa));
    sa.sa_handler = handler;
    sigaction (SIGINT, &sa, 0);
    sigaction (SIGTERM, &sa, 0);
    sa.sa_flags = SA_RESTART;
    sigaction (SIGHUP, &sa, NULL);
#endif

    memset (&sin, 0, sizeof (sin));
    sin.sin_port = htons (port);
    sin.sin_family = AF_INET;
    sin.sin_addr.s_addr = iface;

    s = socket (PF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (s < 0)
    {
        perror ("socket");
        exit (1);
    }
    if (bind (s, (struct sockaddr *) &sin, sizeof (sin)) < 0)
    {
        perror ("bind");
        exit (1);
    }
    if (listen (s, 50) < 0)
    {
        perror ("listen");
        exit (1);
    }

#ifndef WIN32
    if (background && fork ())
        _exit (0);
#endif

#if HAVE_SYSLOG
    if (sys_log)
        openlog ("metaserver", LOG_PID, LOG_LOCAL6);
#endif

    location = 0;
#if HAVE_POLL
    memset (&ufd, 0, sizeof (ufd));
    ufd.fd = s;
    ufd.events = POLLIN | POLLHUP;
#endif /* HAVE_POLL */
    for (;;)
    {
#if HAVE_POLL
        i = poll (&ufd, 1, -1);
#else
        FD_ZERO (&set);
        FD_SET (s, &set);
        i = select (s + 1, &set, 0, 0, 0);
#endif /* HAVE_POLL */
        if (i == -1)
        {
            /* re-read configuration files */
            if (sig_meta)
            {
                for (i = 0; i < numhosts; i++)
                    free (hosts[i]);
                numhosts = 0;
                location = 0;
                sig_meta = 0;
                read_metafile (metafile);
                continue;
            }
            perror (
#if HAVE_POLL
                "poll"
#else
                "select"
#endif
                );
            break;
        }

        sinsize = sizeof (sin);
        f = accept (s, (struct sockaddr *) &sin, &sinsize);
        if (f < 0)
        {
            perror ("accept");
            break;
        }
        WRITE (f, hosts[location], strlen (hosts[location]));
        WRITE (f, "\n", 1);
#if HAVE_SYSLOG
        if (sys_log)
            syslog (LOG_ERR, "Redirecting %s to %s", inet_ntoa (sin.sin_addr),
            hosts[location]);
#endif
        location = (location + 1) % numhosts;
        CLOSE (f);
    }
    CLOSE (s);
#if HAVE_SYSLOG
    closelog ();
#endif
    exit (0);
}
